apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.YCSB/Program.cs

```diff
@@ -0,0 +1,513 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax;
+using Google.Cloud.Spanner.V1.Internal.Logging;
+
+namespace Google.Cloud.Spanner.Data.YCSB
+{
+    internal class Program
+    {
+        private static readonly string s_appVersion = "v1.0";
+
+        private static readonly string s_tableName = "table";
+        private static readonly string s_cloudspanner_instanceName = "cloudspanner.instance";
+        private static readonly string s_cloudspanner_databaseName = "cloudspanner.database";
+        private static readonly string s_cloudspanner_projectName = "cloudspanner.project";
+        private static readonly string s_client_type = "client_type";
+        private static readonly string s_num_worker = "num_worker";
+        private static readonly string s_operationcount = "operationcount";
+        private static readonly string s_boundedStaleness = "cloudspanner.boundedstaleness";
+        private static readonly string s_numchannels = "cloudspanner.numchannels";
+        private static readonly string s_prewarm = "cloudspanner.prewarm";
+        private static readonly string s_targetQps = "cloudspanner.qpstarget";
+        private static readonly string s_qpstestTime = "cloudspanner.qpstesttime";
+
+        private static long s_operationsRemaining;
+        private static long s_totaloperations;
+        private static long s_boundedStalenessSeconds;
+        private static readonly List<string> s_keys = new List<string>();
+        private static string s_templateFile;
+
+        private static readonly ConcurrentDictionary<string, object> s_options =
+            new ConcurrentDictionary<string, object>();
+
+        private static bool s_debugMode;
+        private static readonly Random s_globalRandom = new Random(Environment.TickCount);
+
+        private static readonly ThreadLocal<Random> s_rand = new ThreadLocal<Random>(() =>
```

Comment:
I used to recommend this approach to RNGs, but it doesn't produce very random results. See https://stackoverflow.com/questions/36376888/seeding-multiple-random-number-generators

If the randomness doesn't matter too much, it's fine to keep this - otherwise, I'd probably use a single instance and just lock each time. It's annoying, but such is life :(

Reply:
The lack of real randomness is actually OK and I'd rather not introduce any concurrency locks if it doesnt relate to a real world requirement.  This test is extremely perf sensitive